Reconstruction/CRDEcalRec/src/Algorithm/GlobalClusteringAlg.cpp

```diff
@@ -52,6 +69,270 @@ StatusCode GlobalClusteringAlg::ClearAlgorithm(){
+					((m_bar->getBar()==1 || m_bar->getBar()==2)  && m_bar->getPart()==m_1dcluster[i1d]->getPart() && abs(m_bar->getStave()-m_1dcluster[i1d]->getStave())<=1)
+					)
+					{
+						return true;
+					}
+				}
+				else if(m_1dcluster[i1d]->getBar()==(m_zbarnumber-2*(m_bar->getDlayer()-1)))
+				{
+					if(
+					(m_bar->getBar()==1 && m_bar->getPart()==(m_1dcluster[i1d]->getPart()+1) && abs(m_bar->getStave()-m_1dcluster[i1d]->getStave())<=1)||
+					((m_bar->getBar()==(m_zbarnumber-2*(m_bar->getDlayer()-1)) || m_bar->getBar()==((m_zbarnumber-2*(m_bar->getDlayer()-1))-1))  && m_bar->getPart()==m_1dcluster[i1d]->getPart() && abs(m_bar->getStave()-m_1dcluster[i1d]->getStave())<=1)
+					)
+					{
+						return true;
+					}
+				}
+				else
+				{
+					if(abs(m_bar->getBar()-m_1dcluster[i1d]->getBar())<=1 && m_bar->getPart()==m_1dcluster[i1d]->getPart() && abs(m_bar->getStave()-m_1dcluster[i1d]->getStave())<=1)
+					{
+						return true;
+					}
+				}
+			}
+		}		
+	}
+
+	
+	return false;
+}
+
+//1d->2d
+bool GlobalClusteringAlg::ifAdjacent(PandoraPlus::Calo1DCluster* m_1dcluster, PandoraPlus::Calo2DCluster* m_2dcluster)
+{
+	if(m_1dcluster->getBars().at(0)->getDlayer() == m_2dcluster->getCluster().at(0)->getCluster().at(0)->getDlayer())
+	{
+		return ifSameTower(m_1dcluster,m_2dcluster);
+	}
+	else
+	{
+		return false;
+	}
+}
+
+//2d->3d
+bool GlobalClusteringAlg::ifAdjacent(PandoraPlus::Calo2DCluster* m_2dcluster, PandoraPlus::Calo3DCluster* m_3dcluster)
+{
+	if(ifSameTower(m_2dcluster,m_3dcluster))
+	{
+		return true;
+	}
+	
+	std::vector<const PandoraPlus::CaloBar*> bars_2d = m_2dcluster->getBars();
+	std::vector<const PandoraPlus::CaloBar*> bars_3d = m_3dcluster->getBars();
+	
+	for(int i=0; i<bars_2d.size(); i++)
+	{
+		const PandoraPlus::CaloBar* bob = bars_2d.at(i);
+		
+		for(int j=0; j<bars_3d.size(); j++)
+		{
+			const PandoraPlus::CaloBar* alice = bars_3d.at(j);
+			
+			if(ifModuleAdjacent(bob,alice))
+			{
+				return true;
+			}
+		}
+	}
+	
+	return false;
+}
+
+//some other templates and functions
+template<typename T1, typename T2> bool GlobalClusteringAlg::ifSameTower(T1* m_uncluster, T2* m_incluster) //1d-->2d and 2d-->3d(normal situation)
+{
+	std::vector<int> untower = m_uncluster->getTowerID();
+	std::vector<int> intower = m_incluster->getTowerID();
+    for(int i=0; i<untower.size(); i++)
+    {
+        for(int j=0; j<intower.size(); j++)
+        {
+            if(untower.at(i) == intower.at(j))
+            {
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+bool GlobalClusteringAlg::ifModuleAdjacent(const PandoraPlus::CaloBar* bar_2d, const PandoraPlus::CaloBar* bar_3d)
+{
+	PandoraPlus::CaloBar* bob = new PandoraPlus::CaloBar(); //just first layer
+	PandoraPlus::CaloBar* alice = new PandoraPlus::CaloBar(); //second new
+	
+	if(bar_2d->getModule()==m_module && bar_3d->getModule()==1)
+	{
+		bob = bar_2d->Clone();
```

Comment:
Why clone a bar here? If you don't need to modify the bar then don't clone it. 
Clone() will require a new memory space.